java/src/com/ibm/streamsx/topology/generator/spl/SPLGenerator.java

```diff
@@ -83,6 +98,406 @@ public String generateSPL(JsonObject graph) throws IOException {
         return sb.toString();
     }
     
+    private void separateIntoComposites(JsonObject graph){
+
+        compOperatorStarts.add(OpProperties.PARALLEL);
+        compStarts.add(BVirtualMarker.PARALLEL.kind());
+        compEnds.add(BVirtualMarker.END_PARALLEL.kind());
+        
+//        compOperatorStarts.add(null);
+//        compStarts.add(BVirtualMarker.LOW_LATENCY.kind());
+//        compEnds.add(BVirtualMarker.END_LOW_LATENCY.kind());
+        
+
+        // Find composites until there are no more to find.
+        while(createComposite(graph)){
+
+        }     
+    }
+    
+    /**
```

Comment:
I think this method is the core of the system so it should have an overview (as much desscription as possible) how you are finding and creating composites, so we can understand the algorithm later when it has gone out of our current memory.

I think renaming this method would help, because it's not creating a composite I think it's finding and creating a single composite, the one that is the first one discovered at the most nested level. Sort of `findAndCreateMostNestedComposite`.

I saw this as the methods it calls, e.g.  `findCompositeOpsOfAType` have no descriptions of what they do, and possibly the names are misleading. I think  `findCompositeOpsOfAType` is finding a single composite and its operators, not finding multiple composite operators, as its name could be interpreted as.

Reply:
I like `findAndCreateMostNestedComposite`. I'm sure that was was an off-the-cuff name, but it's accurate.

Reply:
Can we just clarify the terms in the description. It's understandable but only with prior knowledge of what is going on.

The issue is with "composite", it's used to mean multiple items.

Eg. _composites are non-overlapping, contiguous subsections of graph_

Well no, the graph has region of operator invocations that this algorithm extracts into a _composite definition_ and replaces the region with an _composite invocation_ of that newly defined composite.

_composites are non-overlapping, contiguous subsections of graph_

**regions** are non-overlapping, contiguous subsections of graph

_an "innermost" composite is a composite that does not contain another composite_

I'd say it's is actually: an innermost region is a region that does not contain another region. 

For a new reader the phrase "not contain another composite" might be taken as the region cannot contain an invocation of a composite.

Basically, anywhere composite it used it needs to be replaced with one of:
 * region
 * composite definition
 * composite invocation